wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift

```diff
@@ -117,124 +111,120 @@ public class MessageSender {
-                        for: message
-                    )
-                }
-
-                WireLogger.messaging.debug("got 412, will retry: \(shouldRetry)")
-                return .success(shouldRetry)
-            }
+    private func handleProteusFailure(message: any ProteusMessage, _ failure: NetworkError) -> Swift.Result<Set<QualifiedClientID>, MessageSendError> {
+        switch failure {
+        case .missingClients(let messageSendingStatus, _):
+            let missingClients = processor.updateClientsChanges(
+                from: messageSendingStatus,
+                for: message
+            ).flatMap(\.value)
+            let qualifiedClientIDs = missingClients.compactMap({ $0.qualifiedClientID })
 
-        case 533:
             guard
-                let payload = Payload.ResponseFailure(response, decoder: .defaultDecoder),
-                let data = payload.data
+                missingClients.count == qualifiedClientIDs.count
             else {
-                return .failure(MessageSendError.failedToParseResponse)
+                return .failure(MessageSendError.missingPrecondition)
             }
+            return .success(Set(qualifiedClientIDs))
+        case .invalidRequestError(let responseFailure, _):
+            switch (responseFailure.code, responseFailure.label) {
+            case (533, _):
+                guard
+                    let data = responseFailure.data
+                else {
+                    return .failure(MessageSendError.networkError(failure))
+                }
 
-            switch data.type {
-            case .federation:
-                payload.updateExpirationReason(for: message, with: .federationRemoteError)
-            case .unknown:
-                payload.updateExpirationReason(for: message, with: .unknown)
-            }
+                switch data.type {
+                case .federation:
+                    responseFailure.updateExpirationReason(for: message, with: .federationRemoteError)
+                case .unknown:
+                    responseFailure.updateExpirationReason(for: message, with: .unknown)
+                }
 
-            return .success(false)
-        default:
-            let payload = Payload.ResponseFailure(response, decoder: .defaultDecoder)
-            if payload?.label == .unknownClient {
-                clientRegistrationDelegate.didDetectCurrentClientDeletion()
+                return .failure(MessageSendError.networkError(failure))
+            default:
+                return .failure(MessageSendError.networkError(failure))
             }
-
-            if case .permanentError = response.result {
-                WireLogger.messaging.warn("got \(response.httpStatus), not retrying")
-                return .success(false)
+        default:
+            if case .permanentError = failure.response?.result {
+                return .success(Set()) // FIXME: [jacob] it's dangerous to retry indefinitely like this
```

Comment:
done in another PR?

Reply:
No, I've kept this I'm keeping the existing behaviour but I think we should re-evaluate how we want to approach this.

Reply:
ok, let's have a ticket for it then if we don't solve it now in the parent ticket

Reply:
https://wearezeta.atlassian.net/browse/WPB-5454